### DISCUSSION TITLE
Possibility to add metadata to chunks and return that with the result in json structure

### DISCUSSION THREAD

**Discussion by ewagner70** [HKUDS/LightRAG#2000] [Ideas]
It would be great to add metadata (page#, document-title, etc.) to the chunks and retrieve them along with the answer in json format for reference purposes. Would be a great check for grounding + actually a classic RAG functionality to provide references (especially for larger set of documents) along with the answers for follow-up actions and documentation.

**Comment by Arslan-Mehmood1** [comment]
+1

**Comment by jbartot** [comment]
+1

**Comment by TZhanabekov** [comment]
+1

**Comment by Aillian** [comment]
any updates regarding this? Very important feature!

**Comment by Arslan-Mehmood1** [comment]
> any updates regarding this? Very important feature!

+1

**Comment by ArindamRoy23** [comment]
+1

**Comment by wpereira22** [comment]
+1

**Comment by emgiezet** [comment]
+1

**Comment by RolphH** [comment]
+1 @YanSte any particular reason to not plan this? Could something already be achieved (adding additional metadata)?

**Comment by corvusMidnight** [comment]
+1

**Comment by Aillian** [comment]
The absence of this feature is the single reason that prevents me from using lightrag...

**Comment by corvusMidnight** [comment]
@YanSte is it possible you changed your mind about this? Or maybe provide the rationale on why that is the case you do not want to implement the feature?

**Comment by corvusMidnight** [comment]
> The absence of this feature is the single reason that prevents me from using lightrag...

Do you have any alternatives that you are currently using?

**Comment by DartAthos** [comment]
+1

**Comment by andytancgi** [comment]
+1

**Comment by thiborose** [comment]
+1

**Comment by RolphH** [comment]
> +1 [@YanSte](https://github.com/YanSte) any particular reason to not plan this? Could something already be achieved (adding additional metadata)?

@danielaskdd could you shed your vision on this?

**Comment by danielaskdd** [comment]
While it's a valuable enhancement, LightRAG has more pressing and critical tasks that require our immediate attention. The method of delivering the saved meta-content to the query initiator needs extensive discussion and clear definition before implementation.

**Reply by ewagner70** [reply]
> While it's a valuable enhancement, LightRAG has more pressing and critical tasks that require our immediate attention. The method of delivering the saved meta-content to the query initiator needs extensive discussion and clear definition before implementation.

@danielaskdd : may I ask on why this (relatively basic but ultimately important) functionality is so difficult to add? furthermore, can you shed some light on your mentioned "more pressing and critical tasks"?

**Comment by danielaskdd** [comment]
Ensuring source attribution for query results is a comprehensive engineering effort. It encompasses a range of tasks including the recognition of layouts, chapters, images, and tables within the source documents, as well as their subsequent preservation and retrieval. Adding metadata into chunks represents only a minor component of this entire source attribution system.

Key ongoing developments for LightRAG include:
- Entity Disambiguation and Consolidation: Merging entities, including synonyms, to improve semantic consistency.
- Raganything & LightRAG Integration: Integrating the Raganything framework into LightRAG for enhanced capabilities.
- Configurable Processing Pipelines: Offering a modular selection of document processing components, including various document extractors, chunking strategies, prompt templates, and LLMs.
- Multi-Workspace Support: Implementing native support for multiple workspaces within the LightRAG Server.

Upon the completion of these initiatives, our next major milestone will be the implementation of robust source attribution for query results.

We welcome community members who are interested and possess the relevant skills to contribute to the LightRAG project. We encourage you to align your contributions with the development priorities outlined in this roadmap.

**Reply by devgrctool** [reply]
New Version Expetcted Date ?

**Comment by danielaskdd** [comment]
PR #2100, A new API endpoint, /query/data, designed to return raw retrieval data from the RAG process without LLM generation. This feature is crucial for data analysis, and use cases that require direct access to knowledge graph retrieval results (entities, relationships, text chunks).

**Comment by azman-nd** [comment]
+1

**Comment by devgrctool** [comment]
I want to create multiple stores through an API, and each store should have its own knowledge graph. Retrieval should also happen using the store ID or name. Is that possible?

**Comment by steinhauers** [comment]
In my opinion, this idea is also suitable for setting up an RBAC system.

**Comment by johnshearing** [comment]
I am pretty sure that all the work required for citing sources can be done by preprocessing the documents before they are given to LightRAG for indexing. So this doesn't seem like a problem for the LightRAG develoment team.
Rather it is a problem for the users.

I work mostly with videos and LightRAG is able to cite all the metadata like video title, video channel, video upload date, and video URL, when it responds to my queries.  This information is extracted from YouTube into a json file when processing the audio to create a transcript. Then rag.ainsert_custom_kg is used to insert these into the index.

It also tells who is speaking and cites the timestamp and provides a link to the source video queued up to the moment where the information is sourced. The trick is add the labels for who is speaking and the time stamps for each line in the transcript before the transcript is indexed by LightRAG.

I have scripts that do all this work before LightRAG ever sees the transcripts. 
I mention this because it seems to me that simple scripts could be written by LightRAG users that do exactly the same thing with regular documents for citing sources.

Please let me know if I have missed something.

**Comment by RubenGarcia** [comment]
+1

**Comment by bibin-aplxs** [comment]
+1
